src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs

```diff
@@ -675,28 +659,109 @@ void writeChildrenProperty(AbstractNode type, List<Property> publicIOperationPro
+                        Brace();
+                        WriteLine("switch (previousSlot)");
+                        Brace();
+
+                        int slot = 0;
+                        for (; slot < orderedProperties.Count; slot++)
+                        {
+                            // Operation.Enumerator starts indexes at -1. For a given property, the general psuedocode is:
+
+                            // case previousSlot:
+                            //     if (element i is valid) return (true, i, 0);
+                            //     else goto i;
+
+                            // If i is an IOperation, is valid means not null. If i is an ImmutableArray, it means not empty.
+                            // As IOperation is fully nullable-enabled, and the abstract `Current` method is nullable, we'll
+                            // get a warning if it attempts to return a null IOperation from such an array, so we don't need
+                            // to have explicit Debug.Assert code for this.
+
+                            // Then, if the property is an immutable array:
+                            // case i when previousIndex + 1 < property[i].Length:
+                            //    return (true, i, previousIndex + 1);
+
+                            // While the next index is still valid, this will hit this case for i, only moving to the next
+                            // element after the array is exhausted.
+
+                            var previousSlot = slot - 1;
+                            var prop = orderedProperties[slot];
+
+                            WriteLine($"case {previousSlot}:");
+                            Indent();
+
+                            bool isImmutableArray = IsImmutableArray(prop.Type, out _);
+                            if (isImmutableArray)
+                            {
+                                WriteLine($"if (!{prop.Name}.IsEmpty) return (true, {slot}, 0);");
+                            }
+                            else
+                            {
+                                WriteLine($"if ({prop.Name} != null) return (true, {slot}, 0);");
+                            }
+
+                            WriteLine($"else goto case {slot};");
+
+                            Outdent();
+
+                            if (isImmutableArray)
+                            {
+                                WriteLine($"case {slot} when previousIndex + 1 < {prop.Name}.Length:");
+                                Indent();
+                                WriteLine($"return (true, {slot}, previousIndex + 1);");
+                                Outdent();
+                            }
+                        }
+
+                        WriteLine($"case {slot - 1}:");
+                        Indent();
+                        WriteLine("return (false, int.MinValue, int.MinValue);");
```

Comment:
>"return (false, int.MinValue, int.MinValue); [](start = 34, length = 44)

Is this going to cause a throw on a subsequent MoveNext call? This would violate the contract: "If MoveNext passes the end of the collection, the enumerator is positioned after the last element in the collection and MoveNext returns false. When the enumerator is at this position, subsequent calls to MoveNext also return false until Reset is called." I think once we reach "eof" slot, we should remain in this slot.

Reply:
I think something like ```$"return (false, {slot - 1}, {amount of items in the last property});"``` Alternatively, and perhaps simpler, for the case when the last property is an array, would be to "introduce" another case to handle "eof" - ``` $"case {slot}:"```. Both cases (```$"case {slot - 1}"``` and ```$"case {slot}:"``) would ```$"return (false, {slot}, 0);"```. Or ```$"case {slot - 1}"``` could simply $"goto case {slot};"


---
In reply to: [541126952](https://github.com/dotnet/roslyn/pull/49804#discussion_r541126952) [](ancestors = 541126952)

Reply:
I was thinking that this would be as simple as `return (false, previousSlot, previousIndex)`

---
In reply to: [541146522](https://github.com/dotnet/roslyn/pull/49804#discussion_r541146522) [](ancestors = 541146522,541126952)

Reply:
We want ```GetCurrent``` to throw on eof

---
In reply to: [541149260](https://github.com/dotnet/roslyn/pull/49804#discussion_r541149260) [](ancestors = 541149260,541146522,541126952)

Reply:
After looking at two distinct scenarios that we need to handle, I think an extra slot is needed in both cases in order for GetCurrent to work properly as well. And the code we need to generate is always the same:
```
case {slot - 1}:
    goto case {slot};
case {slot}:
    return (false, {slot}, 0);
default:
    throw ExceptionUtilities.UnexpectedValue((previousSlot, previousIndex));
``` 

---
In reply to: [541152163](https://github.com/dotnet/roslyn/pull/49804#discussion_r541152163) [](ancestors = 541152163,541149260,541146522,541126952)